crates/ecs/src/systems/mod.rs

```diff
@@ -607,61 +802,110 @@ macro_rules! impl_system_parameter_function {
+                    Vec<ArchetypeIndex>,
+                );
+
+                fn borrow<'world>(
+                    world: &'world World,
+                    _: &[ArchetypeIndex],
+                ) -> SystemParameterResult<Self::BorrowedData<'world>> {
+                    let archetypes = <($([<P$parameter>],)*) as SystemParameters>::get_archetype_indices(world);
+
+                    Ok((
+                        ($([<P$parameter>]::borrow(world, &archetypes)?,)*),
+                        world,
+                        archetypes,
+                    ))
                 }
-            }
 
-            impl<'a, $([<P$parameter>]: SystemParameter,)*> Query<'a, ($([<P$parameter>],)*)> {
-                fn try_into_iter(self) -> SystemParameterResult<QueryIterator<'a, ($([<P$parameter>],)*)>> {
-                    let base_signature: Vec<TypeId> = [$([<P$parameter>]::base_signature(),)*]
-                        .into_iter()
-                        .flatten()
-                        .collect();
+                fn split_borrowed_data<'borrowed>(
+                    borrowed: &'borrowed mut Self::BorrowedData<'_>,
+                    segment: FixedSegment,
+                ) -> Vec<Self::SegmentData<'borrowed>> {
+                    match segment {
+                        FixedSegment::Single => {
+                            vec![(
+                                ($([<P$parameter>]::split_borrowed_data(&mut borrowed.0.$parameter, FixedSegment::Single),)*),
+                                borrowed.1,         // world
+                                borrowed.2.clone(), // archetypes
+                            )]
+                        }
+                        FixedSegment::Size { segment_size: _, segment_count } => {
+                            let mut segments = Vec::with_capacity(segment_count);
+
+                            #[allow(trivial_casts)]
+                            $(let [<borrowed_pointer_$parameter>]: *mut _ = &mut borrowed.0.$parameter;)*
+
+                            // SAFETY: When querying on multiple threads, all segments need a mutable reference to borrowed.
+                            // Borrowed is never mutated except when the Write system parameter is used.
+                            // To ensure safety, intra-parallelization is disabled if any system parameter needs write access.
+                            // This check is done using the "support_parallelization" method on the Query system parameter.
+                            for _ in 0..segment_count {
+                                segments.push((
+                                    ($([<P$parameter>]::split_borrowed_data(unsafe { &mut *[<borrowed_pointer_$parameter>] }, FixedSegment::Single),)*),
+                                    borrowed.1,         // world
+                                    borrowed.2.clone(), // archetypes
+                                ));
+                            }
```

Comment:
> Why do read-only segments need a mutable reference?

All `SystemParameter` use the same trait, and `Write` needs a mutable reference, so all `SystemParameter` will get a mutable reference.

> Why is unsafe being used instead of split_at_mut?

All segments need to reference all components, because this is for foreign entity access. `split_at_mut` is used in `split_borrowed_data` for `Write` instead. 

> Why can't two segments write to different parts of the component vec at the same time? "intra-parallelization is disabled if any system parameter needs write access"

Here, "if any system parameter needs write access" is referring the system parameters that are inside the `Query`. `Query<(Write<A>,)>` cannot use intra-parallelization because all threads will reference all `A` components.

> Is only intra-system parallelization of component fetching disabled if any parameter needs write access, or execution as well?

Both component fetching and execution if any parameter in `Query` needs write access. 
```rust
fn system_a(a: Read<A>, query: Query<(Write<B>,)>); // Will run sequentially
fn system_b(mut a: Write<A>, query: Query<(Read<B>,)>); // Will use intra-system parallelization
```

I will clarify this comment.

Reply:
Ah, I see. It seems I had missed that this implementation was specifically for `Query` and thus focused on foreign entity access. I think I imagined that we still use the `Query` for normal system execution as well.

> I will clarify this comment.

Thanks!